src/accountsdb/db.zig

```diff
@@ -2030,6 +1995,78 @@ pub const AccountsDB = struct {
         return biggest;
     }
 
+    /// TODO: a number of these parameters are temporary stand-ins for data that will be derived from the state
+    /// of AccountsDB, which currently doesn't all exist.
+    pub fn writeSnapshotTar(
+        /// Although this is a mutable pointer, this method performs no mutations;
+        /// the mutable reference is simply needed in order to obtain a lock on some
+        /// fields.
+        self: *Self,
+        /// `std.io.GenericWriter(...)` | `std.io.AnyWriter`
+        archive_writer: anytype,
+        /// Temporary: See above TODO
+        status_cache: StatusCache,
+        /// Temporary: See above TODO
+        bank_fields: BankFields,
+        /// Temporary: See above TODO
+        bank_fields_inc: BankFields.Incremental,
+        /// Temporary: See above TODO
+        lamports_per_signature: u64,
+        /// Temporary: See above TODO
+        bank_hash_info: BankHashInfo,
+        /// For tests against older snapshots. Should just be 0 during normal operation.
+        stored_meta_write_version: u64,
+    ) !void {
+        // NOTE: we hold the lock for the entire duration of the function to ensure
+        // flush and clean do not create files while generating a snapshot.
+        const file_map, var file_map_lg = self.file_map.readWithLock();
+        defer file_map_lg.unlock();
+
+        const max_rooted_slot = self.largest_root_slot.load(.unordered);
+        const version = sig.version.CURRENT_CLIENT_VERSION;
+
+        var serializable_file_map = std.AutoArrayHashMap(Slot, AccountFileInfo).init(self.allocator);
+        defer serializable_file_map.deinit();
+        try serializable_file_map.ensureTotalCapacity(file_map.count());
+
+        const is_incremental = bank_fields_inc.snapshot_persistence != null;
+
+        for (file_map.values()) |*account_file_rw| {
+            const account_file, var account_file_lg = account_file_rw.readWithLock();
+            defer account_file_lg.unlock();
+
+            if (is_incremental) {
+                if (account_file.slot != max_rooted_slot) continue;
+            } else {
+                if (account_file.slot > max_rooted_slot) continue;
+            }
+
```

Comment:
one detail about incremental snapshots - they can only be created and paired with an existing full snapshot (something we will need to track) - and the slots which incrementals contain will be (base_slot, highest_rooted_slot] - also when we create a new incremental snapshot we will want to delete older incrementals - and when we create a new full snapshot we will want to delete old full and old incremental snapshots

realizing how this is a fair amount of additions lets leave it to another pr and just revert and keep this pr to only full snapshot creations

Reply:
a26b0119023ec1815e43997fc2f2f90393ae072f